---
title: "Showcase: monozygotic twins"
author: "Jonathan Heiss"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Extended QC: monozygotic twins}
  %\VignetteEngine{knitr::rmarkdown_notangle}
  \usepackage[utf8]{inputenc}
---

<style>
body{ font-size: 12pt; text-align: justify; }
p.caption {text-align: justify; font-style: italic; margin-left: 8mm;}
</style>


```{r, include = FALSE}
is_check <- ("CheckExEnv" %in% search()) || any(c("_R_CHECK_TIMINGS_",
             "_R_CHECK_LICENSE_") %in% names(Sys.getenv()))
knitr::opts_chunk$set(eval = !is_check)
```

```{r libraries, include=FALSE}
knitr::opts_chunk$set(fig.align="center")
```

```{r include=FALSE}
library(stringi)
library(data.table)
library(magrittr)
library(ewastools)
library(purrr)
```

*This vignette demonstrates the use of the ewastools package on a public available 450K dataset, with a focus on the implemented quality checks for identifying failed assays, mislabeled or contaminated samples. ewastools provides also functions for preprocessing (dye bias correction, normalization) and estimation of leukocyte composition.*

Download and parse the metadata for the GSE61496 dataset. This dataset is comprised of 150 pairs of monozygotic twins (i.e. 300 samples + 10 technical replicates).

```{r}
samples = "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE61496&targ=gsm&form=text&view=brief"
samples = readLines(samples)
samples = split(samples,cumsum(samples %like% "^\\^SAMPLE = GSM") )

# Extract GSM accessions
names(samples) = map(samples,1) %>% stri_match_first(regex="GSM\\d+")

# Parse meta data
imap(samples,function(s,acc){
	s = strsplit(s,split=" = ",fixed=TRUE)	
	data.table(gsm=acc,variable=map_chr(s,1),value=map_chr(s,2))
}) -> samples

samples = rbindlist(samples)

# Keep only information on sample characteristics and supplementary files
samples = samples[variable %chin% c("!Sample_characteristics_ch1","!Sample_supplementary_file")]
i = samples[variable == "!Sample_characteristics_ch1",which=TRUE]
ch = samples$value[i] %>% stri_split(fixed=": ")
samples$variable[i] = map_chr(ch,1)
samples$value   [i] = map_chr(ch,2)
rm(ch)

# Find the URLs pointing to the two .idat files
samples[variable=="!Sample_supplementary_file" & value %like% "_Red\\.idat",variable:="red"]
samples[variable=="!Sample_supplementary_file" & value %like% "_Grn\\.idat",variable:="grn"]

samples[variable=="sex, 1=m, 2=f",variable:="sex"]
samples[variable=="pair id",variable:="pair_id"]
samples[variable=="birth-weight",variable:="birth_weight"]

# Reshape data.table from long to wide format
samples = dcast(samples, gsm ~ variable)

samples[,sex:=factor(sex,levels=1:2,labels=c("m","f"))]
samples$birth_weight %<>% as.integer
samples$pair_id %<>% as.integer
samples$age %<>% as.integer
samples$tissue = NULL
samples %<>% na.omit

```

Remove technical replicates, sort by `pair_id`, and select first 20 samples
```{r}
samples %<>% unique(by=c("pair_id","birth_weight"))
samples = samples[order(pair_id,birth_weight)]
samples = samples[1:20]
```

Swap two samples on purpose.
```{r}
swap = samples[4,list(red,grn)]
samples[4,c("red","grn"):= samples[5,list(red,grn)]]
samples[5,c("red","grn"):= swap]
rm(swap)
```

Download the dataset.
```{r}
map2(samples$red, samples$gsm %s+% "_Red.idat.gz", ~ download.file(.x,.y) )
map2(samples$grn, samples$gsm %s+% "_Grn.idat.gz", ~ download.file(.x,.y) )
samples$red = NULL; samples$grn = NULL
```

Read in the dataset.
```{r}
meth = read_idats(samples$gsm,quiet=TRUE)
```

The first check evaluates 17 control metrics. The "Bisulfite Conversion II" metric is exemplary plotted below. 3 samples fall below 1, which is the cut-off recommended by Illumina.
```{r fig.width=6,fig.height=3.5}
ctrls = control_metrics(meth)
stripchart(ctrls$`Bisulfite Conversion II`,method="jitter",pch=4,xlab='Bisulfite Conversion II',xlim=c(0,10))
abline(v=1,col=2,lty=3)
```

A logical vector of passed/failed is returned by `sample_failure()`. In this case, 26 samples were flagged. These samples warrant closer inspection (we'll skip this).
```{r}
sample_failure(ctrls) %>% table
```

Before the next check, dye bias is corrected and beta-values are computed. No normalization is performed.
```{r}
meth %<>% correct_dye_bias
beta = dont_normalize(meth)
```

Sex check. The samples swapped above resulted in sex mismatches plotted in red. The actual sex of a sample donor can be inferred by calling `predict_sex()`.
```{r fig.width=6,fig.height=6}
samples[,c("X","Y") := check_sex(meth)]

samples$predicted_sex = predict_sex(samples$X,samples$Y,which(samples$sex=='m'),which(samples$sex=='f'))

plot(X~Y,data=samples,pch=ifelse(samples$sex=="f",1,4),asp=1,xlab="Normalized X chromosome intensities",ylab="Normalized Y chromosome intensities")
points(X~Y,data=samples[sex!=predicted_sex],pch=ifelse(samples$sex=="f",1,3),col=2)
legend("topright",pch=c(1,4),legend=c("female","male"))
```

Genotype calling and classification of SNP outliers. First extract SNP probe positions from the manifest and use the matrix of beta-values as input for `call_genotypes`. This function estimates the parameters of a mixture model and returns posterior probabilities used as soft classification.
```{r}
snps = meth$manifest[probe_type=='rs',index]
snps = beta[snps,]
snps = call_genotypes(snps,learn=FALSE)
```

Use the classified SNPs to compare genetic fingerprints between samples. Monozygotic twins are in this check treated as one donor. List all conflicts, e.g. unexpected agreement or disagreement of fingerprints. The sample swap resulted in 8 conflicts: rows 1,4 and 5 indicate that sample #65 does not have the same fingerprint as #59, #303 and #309, even though they are supposed to come from the same donor. Analogously, row 6 indicates that sample #145 does not match the genetic fingerprint of sample #151. Rows 2,3,7 and 8 on the other hand shows the unexpected fingerprint matches between samples that are supposed to come from two different donors.
```{r}
check_snp_agreement(snps$gamma,1-snps$outliers,samples$pair_id,samples$gsm)
```

The same data can be shown as a plot. There are four (overlapping) outliers corresponding to the entries in the list above.

```{r fig.width=6, fig.height=4}
ewastools:::agreement_(snps$gamma,1-snps$outliers,samples$pair_id,samples$gsm,names=c("Unrelated","Twins"))
```

Check for outliers among the SNP probes. Such can indicate sample contamination or other technical problems. All 20 samples pass our recommended cut-off.
```{r}
samples$log_odds = snp_outliers(snps)
table(samples$log_odds > -4)
```

