<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Recommended Work Flow • ewastools</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Recommended Work Flow">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ewastools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/exemplary_ewas.html">Recommended Work Flow</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Recommended Work Flow</h1>
                        <h4 class="author">Jonathan Heiss</h4>
            
            <h4 class="date">2019-04-22</h4>
      
      
      <div class="hidden name"><code>exemplary_ewas.Rmd</code></div>

    </div>

    
    
<style type="text/css">
body{ font-size: 12pt; text-align: justify; }
img{ border: 0px;}
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
</style>
<p>This vignette exemplifies how to use the <code>ewastools</code> package to clean and pre-process DNA methylation data. After loading the required packages, analysis would start with gathering the phenotype data. In this example using a public dataset from the <a href="https://www.ncbi.nlm.nih.gov/geo/">Gene Expression Omnibus</a> repository, the phenotype data is stored in a file named <code>pheno.csv</code></p>
<pre><code>devtools::install_github("hhhh5/ewastools@master")
library(ewastools)
library(stringi)
library(data.table)
library(magrittr)   
library(purrr)
library(svd)</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pheno =<span class="st"> </span><span class="kw">fread</span>(<span class="st">"pheno.csv"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(pheno)
##           gsm sex smoker
## 1: GSM2260480   m smoker
## 2: GSM2260482   m smoker
## 3: GSM2260485   m smoker
## 4: GSM2260486   m smoker
## 5: GSM2260487   f smoker
## 6: GSM2260488   m smoker</code></pre></div>
<p><code>pheno</code> contains a column <code>gsm</code>, which in this case represents also the prefix of the .idat file names. Usually, however, the file names are a combination of the Sentrix barcode, Sentrix position and color channel and will look something like this <code>200379120004_R01C01_Red.idat</code> for the .idat containing the red color channel, and analogously <code>200379120004_R01C01_Grn.idat</code> for the .idat containing the green color channel. <code>read_idats</code> can be used to import methylation data. It’s first argument is a character vector containing the absolute or relative file paths and names but without the color channel and file extension, e.g. <code>C:/folder/subfolder/200379120004_R01C01</code>. Both red and green .idat files of a particular sample need to be in the same folder.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">meth =<span class="st"> </span><span class="kw"><a href="../reference/read_idats.html">read_idats</a></span>(pheno<span class="op">$</span>gsm,<span class="dt">quiet=</span><span class="ot">TRUE</span>) <span class="co"># `quiet=TRUE` supresses the progress bar</span>
## [1] 622399</code></pre></div>
<p>The entire pre-processing, including filtering by detection p-values, dye-bias correction and conversion into beta-values, can be done in one line …</p>
<pre><code>beta = meth %&gt;% detectionP %&gt;% mask(0.01) %&gt;% correct_dye_bias %&gt;% dont_normalize</code></pre>
<p>… but we will break it up in order to describe the various steps.</p>
<p>The first step should be to filter out unreliable data points which result from low fluorescence intensities. These can be the result of insufficiently amplified DNA. Filtering is done using so-called <strong>detection p-values</strong>, calculated from comparing fluorescence intensities to a noise distribution. Probes below a chosen significance threshold are deemed detected, otherwise undetected. The conventional way of calculating these p-values, as implemented in the GenomeStudio software, lets many unreliable data points pass, demonstrated by the fact that many probes targeting the Y chromosome are classified as detected. <code>ewastools</code> implements an improved estimation of noise levels that improves accuracy.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">meth <span class="op">%&lt;&gt;%</span><span class="st"> </span>detectionP
P.new =<span class="st"> </span>meth<span class="op">$</span>detP</code></pre></div>
<p>For easy comparison a function <code><a href="../reference/detectionP.html">detectionP.neg()</a></code> is provided, which estimates background the conventional way.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">P.neg =<span class="st"> </span>meth <span class="op">%&gt;%</span><span class="st"> </span>detectionP.neg <span class="op">%$%</span><span class="st"> </span>detP</code></pre></div>
<p>We can see the improved accuracy by counting the number of Y chromosome probes that are called detected for a male and a female samples.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">chrY =<span class="st"> </span>meth<span class="op">$</span>manifest[chr<span class="op">==</span><span class="st">'Y'</span>,index]

  male =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/which">which</a></span>(pheno<span class="op">$</span>sex<span class="op">==</span><span class="st">"m"</span>)[<span class="dv">1</span>]
female =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/which">which</a></span>(pheno<span class="op">$</span>sex<span class="op">==</span><span class="st">"f"</span>)[<span class="dv">1</span>]

P.neg =<span class="st"> </span>P.neg[chrY,<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(male,female)]
P.new =<span class="st"> </span>P.new[chrY,<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(male,female)]

P.neg =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">colSums</a></span>(P.neg<span class="op">&lt;</span><span class="fl">0.01</span>,<span class="dt">na.rm=</span><span class="ot">TRUE</span>)
P.new =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">colSums</a></span>(P.new<span class="op">&lt;</span><span class="fl">0.01</span>,<span class="dt">na.rm=</span><span class="ot">TRUE</span>)

<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(P.neg) =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"male"</span>,<span class="st">"female"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(P.new) =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"male"</span>,<span class="st">"female"</span>)</code></pre></div>
<p>Using the conventional detection p-value, for the female sample 402 Y chromosome probes are called detected, a number close to all 416 Y chromosome probes as for the male sample.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">P.neg
##   male female 
##    416    402</code></pre></div>
<p>Using the improved method gives a much more accurate result with all 416 Y chromosome probes classified as detected for the male sample, but only 402 probes classified as detected for the female sample. More information can be found in <a href="https://doi.org/10.1186/s13148-019-0615-3">Heiss and Just, 2019</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">P.new
##   male female 
##    416     73</code></pre></div>
<p>We used a significance threshold of 0.01 above. Moving forward, probes above this threshold should be masked, i.e. set to missing.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">meth <span class="op">%&lt;&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/detectionP.html">mask</a></span>(<span class="fl">0.01</span>)</code></pre></div>
<p>Infinium BeadChips use two fluorescent dyes that are linked to the nucleotides used in the the single-base extension step. A and T nucleotides use are linked with a red dye (the red color channel), G and C nucleotides are linked with a green dye (green color channel). Uncorrected data usually feature higher intensities in the red color channel, the so-called dye bias. For probes of Infinium type II design, which use separate color channels to measure the methylated and unmethylated signal, this results in a shifted distribution of beta-values. (Probes of Infinium design type I are not affected, as they measure both signals in the same color channel.) Dye-bias correction normalizes the red and green color channel. <code>ewastools</code> provides an improved version of RELIC (<a href="https://doi.org/10.1186/s12864-016-3426-3">Xu et al., 2017</a>) using robust Theil-Sen estimators.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">meth <span class="op">%&lt;&gt;%</span><span class="st"> </span>correct_dye_bias</code></pre></div>
<p>The final step is the conversion of intensities to beta-values. While <code>ewastools</code> implements the LOESS normalization <a href="https://doi.org/10.3389/fgene.2015.00092">(Heiss and Brenner, 2015)</a>, I advise against normalization as it does little to protect against batch effects but can result in the removal of genuine biological signal. Instead I recommended to adjust for relevant technical covariates in regression models later.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">beta =<span class="st"> </span><span class="kw"><a href="../reference/Preprocessing.html">dont_normalize</a></span>(meth)</code></pre></div>
<p>Before beginning with the actual epigenome-wide association study, it is advised to check a dataset for problematic samples.</p>
<div id="quality-checks" class="section level2">
<h2 class="hasAnchor">
<a href="#quality-checks" class="anchor"></a>Quality checks</h2>
<p><strong>Control metrics</strong></p>
<p>The first quality check evaluates 17 control metrics which are describe in the <a href="https://support.illumina.com/content/dam/illumina-support/documents/documentation/chemistry_documentation/infinium_assays/infinium_hd_methylation/beadarray-controls-reporter-user-guide-1000000004009-00.pdf">BeadArray Controls Reporter Software Guide</a> from Illumina. Exemplary, the “Bisulfite Conversion II” metric is plotted below. Three samples fall below the Illumina-recommended cut-off of 1. Input for <code><a href="../reference/control_metrics.html">control_metrics()</a></code> is the output of <code><a href="../reference/read_idats.html">read_idats()</a></code>, e.g. the object holding raw or dye-bias-corrected intensities.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ctrls =<span class="st"> </span><span class="kw"><a href="../reference/control_metrics.html">control_metrics</a></span>(meth)
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/stripchart">stripchart</a></span>(ctrls<span class="op">$</span><span class="st">`</span><span class="dt">Bisulfite Conversion II</span><span class="st">`</span>,<span class="dt">method=</span><span class="st">"jitter"</span>,<span class="dt">pch=</span><span class="dv">4</span>,<span class="dt">xlab=</span><span class="st">'Bisulfite Conversion II'</span>,<span class="dt">xlim=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">0</span>,<span class="dv">10</span>))
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/abline">abline</a></span>(<span class="dt">v=</span><span class="dv">1</span>,<span class="dt">col=</span><span class="dv">2</span>,<span class="dt">lty=</span><span class="dv">3</span>)</code></pre></div>
<p><img src="exemplary_ewas_files/figure-html/unnamed-chunk-12-1.png" width="576" style="display: block; margin: auto;"></p>
<p>A logical vector of passed/failed is returned by <code><a href="../reference/control_metrics.html">sample_failure()</a></code> which compares all 17 metrics against the thresholds recommended by Illumina. In this case all samples passed (i.e., <code>failed==FALSE</code>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pheno<span class="op">$</span>failed =<span class="st"> </span><span class="kw"><a href="../reference/control_metrics.html">sample_failure</a></span>(ctrls)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/table">table</a></span>(pheno<span class="op">$</span>failed)
## 
## FALSE 
##    35</code></pre></div>
<hr>
<p><strong>Sex mismatches</strong></p>
<p>The sex of the sample donor can reliable be inferred from the methylation data. This functionality is implemented by the combination of <code><a href="../reference/check_sex.html">check_sex()</a></code> and <code><a href="../reference/check_sex.html">predict_sex()</a></code>. <code><a href="../reference/check_sex.html">check_sex()</a></code> computes the normalized average total fluorescence intensities of the probes targeting the X and Y chromosome. <code><a href="../reference/check_sex.html">predict_sex()</a></code> uses the output of <code><a href="../reference/check_sex.html">check_sex()</a></code> and recorded sex in order to identify mislabeled samples. The function <code><a href="../reference/check_sex.html">check_sex()</a></code> should be applied to dye-bias corrected intensities.</p>
<p>Plotted below are the normalized average total fluorescence intensities of X and Y chromosome probes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pheno[,<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"X"</span>,<span class="st">"Y"</span>) <span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="kw"><a href="../reference/check_sex.html">check_sex</a></span>(meth)]
pheno[,predicted_sex<span class="op">:</span><span class="er">=</span><span class="kw"><a href="../reference/check_sex.html">predict_sex</a></span>(X,Y,<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/which">which</a></span>(sex<span class="op">==</span><span class="st">"m"</span>),<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/which">which</a></span>(sex<span class="op">==</span><span class="st">"f"</span>))]

tmp =<span class="st"> </span>pheno[predicted_sex<span class="op">==</span>sex]
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(Y <span class="op">~</span><span class="st"> </span>X,<span class="dt">data=</span>tmp,<span class="dt">pch=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ifelse">ifelse</a></span>(tmp<span class="op">$</span>sex<span class="op">==</span><span class="st">"f"</span>,<span class="dv">1</span>,<span class="dv">4</span>),<span class="dt">asp=</span><span class="dv">1</span>,<span class="dt">xlab=</span><span class="st">"Normalized X chromosome intensities"</span>,<span class="dt">ylab=</span><span class="st">"Normalized Y chromosome intensities"</span>)
tmp =<span class="st"> </span>pheno[predicted_sex<span class="op">!=</span>sex]
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/points">points</a></span>(Y <span class="op">~</span><span class="st"> </span>X,<span class="dt">data=</span>tmp,<span class="dt">pch=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ifelse">ifelse</a></span>(tmp<span class="op">$</span>sex<span class="op">==</span><span class="st">"f"</span>,<span class="dv">1</span>,<span class="dv">4</span>),<span class="dt">col=</span><span class="dv">2</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/legend">legend</a></span>(<span class="st">"topright"</span>,<span class="dt">pch=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">4</span>),<span class="dt">legend=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"female"</span>,<span class="st">"male"</span>))</code></pre></div>
<p><img src="exemplary_ewas_files/figure-html/unnamed-chunk-14-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Samples form two cluster of males (top left) and females (bottom left). The one mislabeled sample here (in red) can easily be identified and should be flagged.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pheno[sex<span class="op">!=</span>predicted_sex,exclude<span class="op">:</span><span class="er">=</span><span class="ot">TRUE</span>] <span class="co"># flag sample</span>
pheno[sex<span class="op">!=</span>predicted_sex,.(gsm,sex,predicted_sex)]
##           gsm sex predicted_sex
## 1: GSM2260573   f             m</code></pre></div>
<p>Another sample falls outside the two clusters.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pheno[X <span class="op">%between%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.85</span>,<span class="fl">0.95</span>) <span class="op">&amp;</span><span class="st"> </span>Y <span class="op">%between%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.65</span>,<span class="fl">0.75</span>),.(gsm,X,Y,sex,predicted_sex)]
##           gsm         X        Y sex predicted_sex
## 1: GSM2260653 0.9172741 0.710014   m             m</code></pre></div>
<p>There are several possible explanations for samples not clustering with males or females, for example chromosome abnormalities. Or sample contamination. The latter theory can be tested in the next quality check.</p>
<hr>
<p><strong>Genotype calling and outliers</strong></p>
<p>For the next check we first need the row indexes of the SNP probes in <code>beta</code>. <code>meth</code>, the output of <code><a href="../reference/read_idats.html">read_idats()</a></code>, contains a data.table object with the relevant information.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">meth<span class="op">$</span>manifest
##           probe_id addressU addressM channel next_base chr  mapinfo strand
##      1: rs10796216 14622465 41635319     Red         A           NA       
##      2:   rs715359 18796328 48710462     Grn         C           NA       
##      3:  rs1040870 22687484 20663453     Red         A           NA       
##      4: rs10936224 34619331 30630453     Red         T           NA       
##      5:   rs213028 10622451 24684377     Red         T           NA       
##     ---                                                                   
## 485573: cg17939569 73757458       NA    Both             Y 27009430      F
## 485574: cg13365400 61745505       NA    Both             Y 27210334      F
## 485575: cg21106100 56793430       NA    Both             Y 28555536      R
## 485576: cg08265308 67794346 26610401     Red         A   Y 28555550      R
## 485577: cg14273923 16749405       NA    Both             Y 28555912      F
##         probe_type  index  OOBi     Ui     Mi
##      1:         rs      1     1  39329 302627
##      2:         rs      2     1  86374 372482
##      3:         rs      3     2 120242  99780
##      4:         rs      4     3 233473 195590
##      5:         rs      5     4   1025 139183
##     ---                                      
## 485573:         cg 485573    NA 610683 610683
## 485574:         cg 485574    NA 499079 499079
## 485575:         cg 485575    NA 452652 452652
## 485576:         cg 485576 89203 559764 155235
## 485577:         cg 485577    NA  64742  64742</code></pre></div>
<p>SNP probes are labelled <code>"rs"</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">snps =<span class="st"> </span>meth<span class="op">$</span>manifest[probe_type<span class="op">==</span><span class="st">"rs"</span>,index]
snps =<span class="st"> </span>beta[snps,]</code></pre></div>
<p>These SNPs are then used as input for <code><a href="../reference/call_genotypes.html">call_genotypes()</a></code>. This function estimates the parameters of a mixture model consisting of three Beta distributions representing the heterozygous and the two homozygous genotypes, and a fourth component, a uniform distribution, representing outliers. The functions returns posterior probabilities used for soft classification. When setting the argument <code>learn=FALSE</code>, a pre-specified mixture model is used. In this case, we use the pre-specified model as the dataset is quite small and maximum likelihood estimation might be unstable.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genotypes =<span class="st"> </span><span class="kw"><a href="../reference/call_genotypes.html">call_genotypes</a></span>(snps,<span class="dt">learn=</span><span class="ot">FALSE</span>)</code></pre></div>
<p><code><a href="../reference/call_genotypes.html">snp_outliers()</a></code> returns the average log odds of belonging to the outlier component across all SNP probes. I recommend to flag samples with a score greater than -4 for exclusion.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pheno<span class="op">$</span>outlier =<span class="st"> </span><span class="kw"><a href="../reference/call_genotypes.html">snp_outliers</a></span>(genotypes)
pheno[outlier <span class="op">&gt;</span><span class="st"> </span><span class="op">-</span><span class="dv">4</span>,.(gsm,X,Y,outlier)]
##           gsm         X        Y    outlier
## 1: GSM2260653 0.9172741 0.710014 -0.8460863
pheno[outlier <span class="op">&gt;</span><span class="st"> </span><span class="op">-</span><span class="dv">4</span>,exclude<span class="op">:</span><span class="er">=</span><span class="ot">TRUE</span>] <span class="co"># flag sample</span></code></pre></div>
<p>The one sample failing this check is the same sample that did not belong to either the male or female cluster in the plot above. This is strong evidence that this sample is indeed contaminated. While SNP outliers can also result from poorly performing assays, the sample passed the first quality check looking at the control metrics, therefore rendering this possibility unlikely. Another cause for a high outlier score is sample degradation (e.g., FFPE samples).</p>
<p>Other useful functions to be mentioned are <code><a href="../reference/check_snp_agreement.html">check_snp_agreement()</a></code> and <code><a href="../reference/check_snp_agreement.html">enumerate_sample_donors()</a></code>. The former checks whether the genotypes of samples supposed to come from the same donor (or from monozygotic twins) do in fact agree; the latter returns unique IDs for unique genotypes and can, for example, be used to find technical replicates in public datasets.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pheno<span class="op">$</span>donor_id =<span class="st"> </span><span class="kw"><a href="../reference/check_snp_agreement.html">enumerate_sample_donors</a></span>(genotypes)

<span class="co"># List duplicates</span>
pheno[,n<span class="op">:</span><span class="er">=</span>.N,by=donor_id]
pheno[n<span class="op">&gt;</span><span class="dv">1</span>,.(gsm,donor_id)]
##           gsm donor_id
## 1: GSM2260485        3
## 2: GSM2260543        3
pheno[gsm<span class="op">==</span><span class="st">"GSM2260543"</span>,exclude<span class="op">:</span><span class="er">=</span><span class="ot">TRUE</span>] <span class="co"># drop duplicate</span></code></pre></div>
<p>Here samples GSM2260485 and GSM2260543 come from the same donor.</p>
<hr>
<p><strong>PCA</strong></p>
<p>Principal component analysis is a popular feature reduction method: it projects high-dimensional data into a lower-dimensional representation while trying to retain as much variability as possible. This is especially useful when either individual features are highly correlated and it is therefore reasonable to summarize them, or when (sometimes subtle) traces of background effects can be found across of large number of features.</p>
<p>We will drop the X and Y chromosome as we would like to find important drivers of methylation beyond sex.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Random">set.seed</a></span>(<span class="dv">982278</span>)
chrXY =<span class="st"> </span>meth<span class="op">$</span>manifest[chr<span class="op">%in%</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"X"</span>,<span class="st">"Y"</span>) <span class="op">&amp;</span><span class="st"> </span>probe_type<span class="op">!=</span><span class="st">"rs"</span>,index]
pcs =<span class="st"> </span>beta[<span class="op">-</span>chrXY,]
pcs =<span class="st"> </span>pcs <span class="op">-</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">rowMeans</a></span>(pcs)
pcs =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/na.fail">na.omit</a></span>(pcs)
pcs =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/t">t</a></span>(pcs)
pcs =<span class="st"> </span><span class="kw">trlan.svd</span>(pcs,<span class="dt">neig=</span><span class="dv">2</span>) <span class="co"># compute the first two principal components</span>
pcs =<span class="st"> </span>pcs<span class="op">$</span>u

pheno<span class="op">$</span>pc1 =<span class="st"> </span>pcs[,<span class="dv">1</span>]
pheno<span class="op">$</span>pc2 =<span class="st"> </span>pcs[,<span class="dv">2</span>]</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(pc2 <span class="op">~</span><span class="st"> </span>pc1,<span class="dt">pch=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ifelse">ifelse</a></span>(sex<span class="op">==</span><span class="st">"f"</span>,<span class="dv">1</span>,<span class="dv">4</span>),pheno,<span class="dt">asp=</span><span class="dv">1</span>,<span class="dt">xlab=</span><span class="st">"PC #1"</span>,<span class="dt">ylab=</span><span class="st">"PC #2"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/legend">legend</a></span>(<span class="st">"topright"</span>,<span class="dt">pch=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">4</span>),<span class="dt">legend=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"female"</span>,<span class="st">"male"</span>))</code></pre></div>
<p><img src="exemplary_ewas_files/figure-html/unnamed-chunk-23-1.png" width="576" style="display: block; margin: auto;"></p>
<p>There is one clear outlier.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pheno[pc1<span class="op">&lt;</span><span class="st"> </span><span class="op">-</span><span class="fl">0.8</span>,exclude<span class="op">:</span><span class="er">=</span><span class="ot">TRUE</span>]
pheno[pc1<span class="op">&lt;</span><span class="st"> </span><span class="op">-</span><span class="fl">0.8</span>,.(gsm,pc1,pc2)]
##           gsm        pc1        pc2
## 1: GSM2219539 -0.9300883 -0.1484162</code></pre></div>
<p>GSM2219539 is actually a lung tissue sample from another GEO dataset (included here for educational purposes). It dominates the first principal component and should be excluded as it otherwise could drastically change the results of downstream analyses.</p>
<p>PCA may be applied iteratively. After excluding samples that manifest as outliers, repeating PCA can give very different principal components.</p>
<hr>
<p><strong>Leukocyte composition</strong></p>
<p>This quality check will only apply in case of blood samples (blood is, however, one of the most commonly studied tissues). The function <code><a href="../reference/estimateLC.html">estimateLC()</a></code> implements the <a href="https://doi.org/10.1186/1471-2105-13-86">Houseman method</a> to predict the leukocyte composition. The user has the choice between various sets of model parameters trained on various reference datasets (see <code><a href="../reference/estimateLC.html">?estimateLC</a></code> for a list of options). The function operates on the matrix of beta-values.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">LC =<span class="st"> </span><span class="kw"><a href="../reference/estimateLC.html">estimateLC</a></span>(beta,<span class="dt">ref=</span><span class="st">"deGoede+Reinius"</span>)
pheno =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">cbind</a></span>(pheno,LC)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Round">round</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(LC),<span class="dv">3</span>)
##       MO     B   CD4   CD8    GR    NK  nRBC
## 1: 0.048 0.007 0.025 0.087 0.821 0.020 0.000
## 2: 0.059 0.000 0.062 0.097 0.746 0.055 0.007
## 3: 0.050 0.009 0.062 0.047 0.799 0.048 0.000
## 4: 0.079 0.000 0.057 0.085 0.744 0.056 0.005
## 5: 0.079 0.004 0.093 0.074 0.727 0.040 0.011
## 6: 0.058 0.004 0.087 0.072 0.724 0.064 0.000</code></pre></div>
<p><code>LC</code> contains estimated proportions for seven cell types (dependent on the chosen reference dataset).</p>
<p>A second foreign sample from another GEO dataset has been hidden in the dataset, consisting of a purified fraction of granulocytes. Plotting <code>GR</code> this sample can easily be spotted.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(pheno<span class="op">$</span>GR<span class="op">*</span><span class="dv">100</span>,<span class="dt">ylab=</span><span class="st">"Granulocyte fraction (%)"</span>)</code></pre></div>
<p><img src="exemplary_ewas_files/figure-html/unnamed-chunk-26-1.png" width="576" style="display: block; margin: auto;"></p>
<p>It is the third to last sample, GSM1185585.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pheno[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/which.min">which.max</a></span>(GR),.(gsm,GR)]
##           gsm       GR
## 1: GSM1185585 1.009805
pheno[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/which.min">which.max</a></span>(GR),exclude<span class="op">:</span><span class="er">=</span><span class="ot">TRUE</span>]</code></pre></div>
<p>The lung sample is also prominent, with an estimated proportion of <code>GR</code> of (not exactly because of numerical issues) zero.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pheno[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/which.min">which.min</a></span>(GR),.(gsm,GR)]
##           gsm         GR
## 1: GSM2219539 0.02205263
pheno[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/which.min">which.min</a></span>(GR),exclude<span class="op">:</span><span class="er">=</span><span class="ot">TRUE</span>]</code></pre></div>
<p>We drop the problematic samples</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">drop  =<span class="st"> </span>pheno[exclude<span class="op">==</span><span class="ot">TRUE</span>,which=<span class="ot">TRUE</span>]
pheno =<span class="st"> </span>pheno[ <span class="op">-</span>drop]
beta  =<span class="st"> </span>beta [,<span class="op">-</span>drop]
meth <span class="op">%&lt;&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/drop_samples.html">drop_samples</a></span>(<span class="dt">j=</span>drop)</code></pre></div>
</div>
<div id="ewas" class="section level2">
<h2 class="hasAnchor">
<a href="#ewas" class="anchor"></a>EWAS</h2>
<p>You’ve cleaned and pre-processed the data, now it is time for the actual EWAS. First it is important to correctly code the variables. <code>smoker</code> and <code>sex</code> are vectors of type <code>character</code>, but should be converted to factors.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pheno =<span class="st"> </span>pheno[,.(
     gsm
    ,<span class="dt">sex =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/factor">factor</a></span>(sex,<span class="dt">levels=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"m"</span>,<span class="st">"f"</span>)) <span class="co"># first level is the reference</span>
    ,<span class="dt">smoker =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/factor">factor</a></span>(smoker,<span class="dt">levels=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"non-smoker"</span>,<span class="st">"smoker"</span>))
    ,GR,MO,B,CD4,CD8,NK,nRBC
)]</code></pre></div>
<p>We want test all CpG sites for their association with smoking. Unfortunately, the phenotype data is very sparse, as it is typcial for public datasets. Aside from smoking, only sex and the estimated proportions of the seven cell types will be including in the model. The following code snippet is optimized for readability not speed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">f =<span class="st"> </span><span class="cf">function</span>(meth){
    m =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/lm">lm</a></span>(meth<span class="op">~</span><span class="dv">1</span><span class="op">+</span>sex<span class="op">+</span>smoker<span class="op">+</span>GR<span class="op">+</span>MO<span class="op">+</span>B<span class="op">+</span>CD4<span class="op">+</span>CD8<span class="op">+</span>NK<span class="op">+</span>nRBC,<span class="dt">data=</span>pheno)
    <span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/coef">coef</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(m))[<span class="st">"smokersmoker"</span>,<span class="dv">4</span>] <span class="co"># extract the p-value for the smoking</span>
}

f =<span class="st"> </span><span class="kw">possibly</span>(f,<span class="dt">otherwise=</span><span class="ot">NA_real_</span>) <span class="co"># catch errors</span>

pvals =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/apply">apply</a></span>(beta,<span class="dv">1</span>,f)</code></pre></div>
<p>We create a <code>data.table</code> holding the p-values and information about the probes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">SMK =<span class="st"> </span><span class="kw">data.table</span>(<span class="dt">probe_id=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(beta),<span class="dt">pval=</span>pvals)
SMK <span class="op">%&lt;&gt;%</span><span class="st"> </span>na.omit
SMK[,qval<span class="op">:</span><span class="er">=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/p.adjust">p.adjust</a></span>(pval,<span class="dt">m=</span><span class="st">"fdr"</span>)]
SMK =<span class="st"> </span>SMK[qval<span class="op">&lt;</span><span class="fl">0.05</span>]
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(SMK)
##      probe_id         pval       qval
## 1: cg21566642 1.440710e-08 0.00699044
## 2: cg01940273 1.196028e-07 0.01450806
## 3: cg05575921 4.776904e-08 0.01158896
## 4: cg21161138 7.968222e-08 0.01288748</code></pre></div>
<p>Two of the significant CpGs are known biomarkers overlapping genes (ALPPL2, AHRR) for which the association with smoking has been validated in several cohorts <a href="https://dx.doi.org/10.1371%2Fjournal.pone.0063812">(Zeilinger et. al., 2013)</a>.</p>
</div>
<div id="final-comments" class="section level2">
<h2 class="hasAnchor">
<a href="#final-comments" class="anchor"></a>Final comments</h2>
<p>Depending on the dataset, many other types of quality checks might be applicable. If you have suggestions or comments regarding <code>ewastools</code>, please send an <a href="mailto:hhhh5@posteo.de">email</a>, or file an issue or submit a pull request on GitHub (<a href="https://github.com/hhhh5/ewastools" class="uri">https://github.com/hhhh5/ewastools</a>).</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#quality-checks">Quality checks</a></li>
      <li><a href="#ewas">EWAS</a></li>
      <li><a href="#final-comments">Final comments</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Jonathan Alexander Heiss.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
